helix-core/src/surround.rs

```diff
@@ -39,13 +39,98 @@ pub fn find_nth_pairs_pos(
     n: usize,
 ) -> Option<(usize, usize)> {
     let (open, close) = get_pair(ch);
-    // find_nth* do not consider current character; +1/-1 to include them
-    let open_pos = search::find_nth_prev(text, open, pos + 1, n, true)?;
-    let close_pos = search::find_nth_next(text, close, pos - 1, n, true)?;
+
+    let (open_pos, close_pos);
+    if open == close {
+        // find_nth* do not consider current character; +1/-1 to include them
+        open_pos = search::find_nth_prev(text, open, pos + 1, n, true)?;
+        close_pos = search::find_nth_next(text, close, pos - 1, n, true)?;
+    } else {
+        open_pos = find_nth_open_pair(text, open, close, pos, n)?;
+        close_pos = find_nth_close_pair(text, open, close, pos, n)?;
+    }
 
     Some((open_pos, close_pos))
 }
 
+fn find_nth_open_pair(
+    text: RopeSlice,
+    open: char,
+    close: char,
+    mut pos: usize,
+    n: usize,
+) -> Option<usize> {
+    let mut chars = text.chars_at(pos + 1);
+
+    // Adjusts pos for the first iteration, and handles the case of the
+    // cursor being *on* the close character which will get falsely stepped over
+    // if not skipped here
+    if chars.prev()? == open {
+        return Some(pos);
+    }
+
+    for _ in 0..n {
+        let mut step_over: usize = 0;
+
+        loop {
+            let c = chars.prev()?;
+            pos = pos.saturating_sub(1);
+
+            // ignore other surround pairs that are enclosed *within* our search scope
+            if c == close {
+                step_over += 1;
+            } else if c == open {
+                if step_over == 0 {
+                    break;
+                }
+
+                step_over = step_over.saturating_sub(1);
+            }
+        }
+    }
+
+    Some(pos)
+}
+
+fn find_nth_close_pair(
+    text: RopeSlice,
+    open: char,
+    close: char,
+    mut pos: usize,
+    n: usize,
+) -> Option<usize> {
+    if pos >= text.len_chars() {
+        return None;
+    }
+
+    let mut chars = text.chars_at(pos);
+
+    if chars.next()? == close {
+        return Some(pos);
+    }
+
+    for _ in 0..n {
+        let mut step_over: usize = 0;
+
+        loop {
+            let c = chars.next()?;
+            pos += 1;
+
+            if c == open {
+                step_over += 1;
+            } else if c == close {
+                if step_over == 0 {
+                    break;
+                }
+
+                step_over = step_over.saturating_sub(1);
+            }
```

Comment:
I tried this with a match statement, and for whatever reason `open` gets mutated to the value of `c` and `step_over` keeps increasing. Am I missing anything here ?

```rust
             match c {
                 open => step_over += 1,
                 close => {
                     if step_over == 0 {
                         break;
                     }

                     step_over = step_over.saturating_sub(1);
                 }
             }

```

Reply:
That's because the match is shadowing `open` and `close` with new variables. It's as if you wrote `let open = c`, which is an irrefutable pattern and you always get in the first arm. By the way if the behavior was the one you expected this wouldn't compile because of missing cases.
When you write:
```rust
let opt = Some(1);
match opt {
    Some(1) => { … }
    _ => { … }
}
```
It's a bit as if you wrote
```rust
if let Some(1) = opt {
    …
} else {
    …
}
```
Somehow the pattern you're trying to use is working with const values, but I advise against because if the const name is removed or name is changed code will still compiles but behavior will be changed.

You could use a match, but using guards:
```rust
_ if c == open => { … }
```
Note that I think it's better the way you did with `if … else if …`.

Reply:
Aha so the value of `c` gets binded to `open` since we're matching by the value of `c` and not the value of `open`. The compiler letting me do it without an `_ => ...` should've given me a clue about what was happening 😅 . Thanks for the detailed explanation !